docs/specs/deployment-wise-sms-gateway.md

```diff
@@ -0,0 +1,89 @@
+
+## New environment variables
+
+The following environment variables are added
+
+```
+SMS_GATEWAY_TWILIO_ACCOUNT_SID=
+SMS_GATEWAY_TWILIO_AUTH_TOKEN=
+SMS_GATEWAY_TWILIO_MESSAGING_SERVICE_SID=
+
+SMS_GATEWAY_NEXMO_API_KEY=
+SMS_GATEWAY_NEXMO_API_SECRET=
+
+SMS_GATEWAY_CUSTOM_URL=
+SMS_GATEWAY_CUSTOM_TIMEOUT=
+
+SMS_GATEWAY_DEFAULT_USE_CONFIG_FROM=environment_variable|authgear.secrets.yaml
+SMS_GATEWAY_DEFAULT_PROVIDER=twilio|nexmo|custom
+```
+
+- `SMS_GATEWAY_{PROVIDER}_*`
+  - The provider specific configs and credentials.
+- `SMS_GATEWAY_DEFAULT_USE_CONFIG_FROM`
+  - The default sms gateway config source. Can be `environment_variable` or `authgear.secrets.yaml`.
+- `SMS_GATEWAY_DEFAULT_PROVIDER`
+  - The default sms provide. Can be `twillio | nexmo | custom`.
+
+## Addition to authgear.yaml
+
+```diff
+ messaging:
+   sms_provider: twillio | nexmo | custom
++  sms_gateway:
++    use_config_from: environment_variable | authgear.secrets.yaml
++    provider: twillio | nexmo | custom
+```
+
+- `messaging.sms_gateway`
+  - The new sms gateway config, capable to specify reading config from environment_variable or project config.
+- `messaging.sms_gateway.use_config_from`
+  - Required when `messaging.sms_gateway` presents.
+  - Can be from `environment_variable` / `authgear.secrets.yaml`.
+- `messaging.sms_gateway.provider`
+  - Can be `twillio | nexmo | custom`.
+  - When `messaging.sms_gateway.use_config_from` is `environment_variable`, it is optional.
+  - When `messaging.sms_gateway.use_config_from` is `authgear.secrets.yaml`, it is required.
+
+## Behavior explained
+
+### Table 1
+
+| `sms_provider` | `sms_gateway` | Behavior                                                                                                                                |
+| -------------- | ------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
+| absent         | absent        | See table 2.                                                                                                                            |
+| absent         | **present**   | Use sms gateway config. See Table 3                                                                                                     |
+| **present**    | absent        | Use `messaging.sms_provider` from `authgear.secrets.yaml`. Read config from `sms.{messaging.sms_provider}` from `authgear.secrets.yaml` |
+| **present**    | **present**   | `messaging.sms_provider` will be ignored. Use sms gateway config, See Table 3                                                           |
+
+### Table 2
+
+When `sms_provider` and `sms_gateway` are absent, the provider config is termined by environment variables.
+
+| `sms_provider` | `DEFAULT_USE_CONFIG_FROM` | `DEFAULT_PROVIDER` | Behavior                                                                                                                                   |
```

Comment:
Since this table talks about when `sms_provider` and `sms_gateway` are absent, the column `sms_provider` is not needed.

Reply:
Updated